kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,48 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks.
+    /// This callback is called when a digest is computed.
+    /// The callback should follow the `Client` `hash_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes written on success
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8]) -> Result<usize, ReturnCode>;
+
+    /// Request the hardware block to generate a Digest
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self);
+
+    /// Stores the returned digest in the memory location specified. This
+    /// is the only way to get the output of the digest.
+    fn get_digest(&self, digest: &mut [u8]) -> ReturnCode;
+
+    /// Clear the keys and any other sensitive data.
+    /// This won't clear the buffers provided to this API, that is up to the
```

Comment:
What buffers? As far as I understand, the user only provides references to buffers which are valid for the individual calls.

Reply:
The user/app allocates buffers with the key, data and digest. These are passed as references to the digest HIL. I'm just making it clear that these won't be cleared and that's up to the user/app to do.

Reply:
Fine. However that is already guaranteed by the type signature (lifetimes) of the HIL. In it's current state the `clear_data` function could never (without resorting to unsafe code) "clear" any buffer that's just been handed in as a reference to a function. For this, the HIL would need to "own" the mutable reference to the buffer (for instance in `get_digest`), but that doesn't happen, as the automatically derived lifetimes will "hand that mutable buffer reference back" after the function execution (please correct me if I'm wrong here).

Reply:
In the updated version we expose `mut` versions of the buffers, so we could clear them. I'm just making sure it's explicit that we won't clear them.